[spmd] by default average gradients for nccl backend

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100073
* #100938
* #100853
* #100852
* __->__ #99964
* #100072

This PR by default average gradient for NCCL backend, this allows
SPMD's data parallel match with DDP/FSDP results.

wanchaol committed:
[spmd] by default average gradients for nccl backend

This PR by default average gradient for NCCL backend, this allows
SPMD's data parallel match with DDP/FSDP results.

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -283,8 +284,20 @@ def _gen_partial_strategy(mesh: DeviceMesh) -> PlacementStrategy:
+    else:
+        # TODO: we should figure out a way to support avg reduction
+        # for non-NCCL backend (i.e. Gloo)
+        reduce_op = c10d.ReduceOp.SUM
```
is there a way to clearly warn users about this difference? Otherwise, it might be surprising that different backend lead to different numerical results. Or maybe, we can just let Gloo crash for now:

https://github.com/pytorch/pytorch/blob/e43918b93a2f595f8a4bbfb9e4d195d3e8746130/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L193

wanchaol committed:
Update on "[spmd] by default average gradients for nccl backend"


This PR by default average gradient for NCCL backend, this allows
SPMD's data parallel match with DDP/FSDP results.

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] by default average gradients for nccl backend"


This PR by default average gradient for NCCL backend, this allows
SPMD's data parallel match with DDP/FSDP results.

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] by default average gradients for nccl backend"


This PR by default average gradient for NCCL backend, this allows
SPMD's data parallel match with DDP/FSDP results.

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] by default average gradients for nccl backend"


This PR by default average gradient for NCCL backend, this allows
SPMD's data parallel match with DDP/FSDP results.

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] by default average gradients for nccl backend"


This PR by default average gradient for NCCL backend, this allows
SPMD's data parallel match with DDP/FSDP results.

[ghstack-poisoned]

mrshenli reviewed (commented):
LGTM
